src/AbstractRole.php

```diff
@@ -103,16 +103,33 @@ public function addChild($child)
      */
     public function setParent($parent)
     {
-        $this->parent = $parent;
-
+        if (null === $this->parents) {
+            $this->parents = [];
+        }
+        if (!in_array($parent, $this->parents)) {
+            $this->parents[] = $parent;
+        }
         return $this;
     }
 
     /**
-     * @return null|RoleInterface
+     * @return null|RoleInterface|array
      */
     public function getParent()
     {
-        return $this->parent;
+        if (1 === count($this->parents)) {
+            return $this->parents[0];
+        }
+        return $this->parents;
+    }
+
+    /**
+     * @param  RoleInterface $parent
+     * @return RoleInterface
+     */
+    public function addParent($parent)
+    {
+        $parent->addChild($this);
```

Comment:
My concern is that we could hit a circular recursion issue, where the parent calls `addParent()` on the child, which then calls `addChild()` on the parent...